narwhal/primary/src/certificate_waiter.rs

```diff
@@ -354,42 +359,65 @@ async fn fetch_certificates_helper(
     network: &P2pNetwork,
     committee: &Committee,
     request: FetchCertificatesRequest,
-) -> FetchCertificatesResponse {
+) -> Option<FetchCertificatesResponse> {
     trace!("Start sending fetch certificates requests");
     // TODO: make this a config parameter.
-    let request_interval = Duration::from_secs(PARALLEL_FETCH_REQUEST_INTERVAL_SECS);
+    let request_interval = PARALLEL_FETCH_REQUEST_INTERVAL_SECS;
     let mut peers: Vec<NetworkPublicKey> = committee
         .others_primaries(name)
         .into_iter()
         .map(|(_, _, network_key)| network_key)
         .collect();
-    loop {
-        // TODO: order by stake weight instead. Possibly shuffle within a few buckets.
-        peers.shuffle(&mut ThreadRng::default());
+    peers.shuffle(&mut ThreadRng::default());
+    let fetch_timeout = PARALLEL_FETCH_REQUEST_INTERVAL_SECS * peers.len().try_into().unwrap()
+        + PARALLEL_FETCH_REQUEST_ADDITIONAL_TIMEOUT;
+    let fetch_callback = async move {
+        // TODO: shuffle by stake weight instead.
+        debug!("Starting to fetch certificates");
         let mut fut = FuturesUnordered::new();
-        for peer in peers.iter() {
-            let network = network.network();
-            let request = request.clone();
-            fut.push(monitored_future!(async move {
-                network.fetch_certificates(peer, request).await
-            }));
+        // Loop until one peer returns with certificates, or no peer does.
+        loop {
+            if let Some(peer) = peers.pop() {
+                let network = network.network();
+                let request = request.clone();
+                fut.push(monitored_future!(async move {
+                    debug!("Sending out fetch requests in parallel.");
```

Comment:
Since we are adding this I would suggest to also include the `peer_id` in the statement so we can trace to who we are sending the request to